### PR TITLE
Implement shared worktree for the tests

### PR DESCRIPTION
**Pull request by thrix**
Implement a shared `tree` directory for running the tests. This is used for `prepare`, `execute` and `finish` steps when running commands and solves the problem with weird working directory when running the tests.

Data is synced in via `rsync` when creating the plan, we omit by default `.git` repository from syncing.

Added two tests to verify that everything works as expected.

Todo:
- [x]  resolve todo items
- [x]  ignore .git when copying
- [x]  make sure all drivers work
- [x]  add more tests
- [x]  make tests pass

Resolves: #423

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

### CODE REVIEW

File: tmt/steps/__init__.py [hunk 1]
```diff
@@ -26,14 +26,15 @@ class Step(tmt.utils.Common):
...
-    def __init__(self, data={}, plan=None, name=None):
+    def __init__(self, data={}, plan=None, name=None, workdir_tree=False):
         """ Initialize and check the step data """
         super().__init__(name=name, parent=plan)
         # Initialize data
         self.plan = plan
         self.data = data
         self._status = None
         self._plugins = []
+        self.workdir_tree = None
```
Comment (thrix) [inline tmt/steps/__init__.py:37]: seems this is not needed

Comment (thrix) [reply tmt/steps/__init__.py:37]: will be fixed in next patchset

File: tmt/steps/__init__.py [hunk 2]
```diff
@@ -57,6 +58,13 @@ def __init__(self, data={}, plan=None, name=None):
             if 'name' not in data and len(self.data) > 1:
                 raise GeneralError(f"Missing '{self}' name in '{self.plan}'.")
 
+        # Create symlink from run tree to the plan tree
+        if self.workdir and workdir_tree:
+            # TODO add dry mode
```
Comment (thrix) [inline tmt/steps/__init__.py:63]: todo item

File: tests/run/worktree/data/testplan.fmf [hunk 3]
```diff
@@ -0,0 +1,8 @@
+prepare:
+    script: echo CHECK > testfile
```
Comment (lukaszachy) [inline tests/run/worktree/data/testplan.fmf:2]: Idea for prepare test extension: `script: grep PASS existing_file` and create existing_file in next to testplan.fmf So we assert you  can find custom script files in prepare step.

Comment (thrix) [reply tests/run/worktree/data/testplan.fmf:2]: make sense, thx

Comment (thrix) [reply tests/run/worktree/data/testplan.fmf]: I believe resolved

**Comment by thrix** [comment]
@psss @lukaszachy @pvalena @FrNecas I believe this is getting ready for initial review.

Simple use case now works for me, need to add a bit more test and check all the tests, but review should be possible

File: tmt/steps/provision/podman.py [hunk 4]
```diff
@@ -128,16 +128,18 @@ def start(self):
...
-        tmt_workdir = self.parent.plan.workdir
-        self.container = 'tmt' + tmt_workdir.replace('/', '-')
+        self.container = 'tmt' + rundir.replace('/', '-')
         self.verbose('name', self.container, 'green')
 
         # Run the container
         self.container_id = self.podman(
             ['run'] + ['--name', self.container,
-            '-v', f'{tmt_workdir}:{tmt_workdir}:Z', '-itd', self.image],
+            '-v', f'{rundir}:{rundir}:Z', '-itd', self.image],
```
Comment (jscotka) [inline tmt/steps/provision/podman.py:142]: when I'm seeing this under podman, that theoretically you mount folder inside podman, and do not use `podman cp` and pull/push operations, maybe there may exist another similar optimizations for virtual machines, e.g.
 - use `sshfs` (probably has to be called on guest) this could be bad. 
 - use `libguestfs` tools to work with VM images and copy files directly from/to image not do it via `ssh(rsync)`
 - map VM discs to local machine e.g. `loopback` device and then mount the discs locally (not sure if possible without sudo)
 - unify copy operation on local containers (VMs) `machinectl copy-to/from`

Bad is, that various types needs various handling how to copy/share files across systems, but it is also current situation.

Take it as an brainstorming, what I'm thinking about :-)

Comment (thrix) [reply tmt/steps/provision/podman.py:142]: @jscotka thanks for the suggestion. 

> use sshfs (probably has to be called on guest) this could be bad.

I have not good experience with this, I would rather avoid it

> use libguestfs tools to work with VM images and copy files directly from/to image not do it via ssh(rsync)

This could work with testcloud, but it will not with minute provisioner. I would rather stick to `rsync` which works in all cases of VM provisioning

> unify copy operation on local containers (VMs) machinectl copy-to/from

I see the same problem as with the one before.

> Bad is, that various types needs various handling how to copy/share files across systems, but it is also current situation.

Well sticking to `rsync` with all VMs and volume mounting to containers, does not seem so bad to me.

> Take it as an brainstorming, what I'm thinking about :-)

I did, thanks for the suggestions, it is good to have this conversation. 
￼

Comment (thrix) [reply tmt/steps/provision/podman.py:142]: @jscotka if you want to file for this an issue, if you think there could be an improvement, pls file a separate issue / RFE. I consider this resolved for this PR.

**Comment by thrix** [comment]
getting closer to fix all tests \o/

File: tmt/steps/execute/internal.py [hunk 5]
```diff
@@ -117,7 +117,6 @@ def go(self):
         for guest in self.step.plan.provision.guests():
 
             # Push workdir to guest and execute tests
-            guest.push()
```
Comment (thrix) [inline tmt/steps/execute/internal.py:120]: I had to move push to guest to earlier stage, so the worktree is available for the prepare step.

File: tests/run/worktree/data/simple/testplan.fmf [hunk 6]
```diff
@@ -0,0 +1,24 @@
+prepare:
+    script: echo CHECK testfile
```
Comment (psss) [inline tests/run/worktree/data/simple/testplan.fmf:2]:
```suggestion
    script: echo CHECK > testfile
```

File: tests/run/worktree/data/simple/testplan.fmf [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    script: grep CHECK testfile
+
+finish:
+    script: grep CHECK testfile
+
+/local:
+    provision:
+        how: local
+
+# /minute:
```
Comment (psss) [inline tests/run/worktree/data/simple/testplan.fmf:14]: Let's add the new test to the `tier 1` set. It will then be executed as part of the `/plans/provision/*` plans.

File: tmt/steps/provision/__init__.py [hunk 8]
```diff
@@ -392,11 +393,12 @@ def push(self, source=None, destination=None, options=None):
         which are '-Rrz --links --safe-links --delete'.
         """
         if options is None:
-            options = "-Rrz --links --safe-links --delete"
+            options = "-rRrz --links --safe-links --delete"
```
Comment (psss) [inline tmt/steps/provision/__init__.py:396]: The `-r` option is already there.

File: tmt/steps/finish/__init__.py [hunk 9]
```diff
@@ -17,6 +17,10 @@ class Finish(tmt.steps.Step):
     successful) so that provisioned systems are not kept running.
     """
 
+    def __init__(self, data, plan):
+        """ Initialize prepare step data """
+        super().__init__(data, plan)
```
Comment (psss) [inline tmt/steps/finish/__init__.py:22]: Why is this necessary if it only executes parent's init?

File: tmt/base.py [hunk 10]
```diff
@@ -980,6 +989,37 @@ def _save_tree(self, tree):
...
+                self.debug(f'Not using git worktree due to: {error}')
+                pass
+
+        # Sync additional data, except .git
+        self.run(
+            f'rsync -arv --exclude .git {tree_root}/ {workdir_tree}',
+            cwd=self.workdir)
+
+    def _git_worktree(self, tree_root):
+        """ Create git worktree under 'tree' directory """
```
Comment (psss) [inline tmt/base.py:1012]: If we are going to use the `worktree` feature, it seems we will need to take care of the cleanup as well. I did some experimenting and after a few attempts I resulted with this:

```
> git branch
* master
+ var-tmp-tmt-run-330
+ var-tmp-tmt-run-331
+ var-tmp-tmt-run-332
```

Plus, if the run directory is removed, the worktree stays there, so the next time you get:

```
Not using git worktree due to: ("Failed to run command 'git -C /tmp/mini worktree add -b var-tmp-tmt-run-332 /var/tmp/tmt/run-332/tree'.", 'git -C /tmp/mini worktree add -b var-tmp-tmt-run-332 /var/tmp/tmt/run-332/tree', 255, "Preparing worktree (new branch 'var-tmp-tmt-run-332')\n", "fatal: A branch named 'var-tmp-tmt-run-332' already exists.\n")
```

Would it make sense to start with just copying files as the first step to fix the problem?

File: tmt/steps/finish/shell.py [hunk 11]
```diff
@@ -60,4 +60,4 @@ def go(self, guest):
         # Execute each script on the guest
         for script in scripts:
             self.verbose('script', script, 'green')
-            guest.execute(script, cwd=self.step.workdir)
+            guest.execute(script, cwd=self.step.plan.run.workdir_tree)
```
Comment (psss) [inline tmt/steps/finish/shell.py:63]: Do we want to share the `tree` directory across multiple plans? This could cause problems once/if we run plans in parallel.

**Review by psss — CHANGES_REQUESTED** [review]
A couple of questions added. One important thing: We should clearly agree on whether the work tree is supposed/allowed to be shared across plans.

File: tmt/steps/finish/shell.py (hunk 11 above)
Comment (thrix) [reply tmt/steps/finish/shell.py:63]: I guess it would be better not to share it ...

File: tmt/base.py (hunk 10 above)
Comment (thrix) [reply tmt/base.py:1012]: we can just cleanup in finish, not a big problem

Comment (thrix) [reply tmt/base.py:1012]: of course, if somebody sees a problem with that, we can just do rsync

File: tmt/utils.py [hunk 12]
```diff
@@ -125,6 +125,10 @@ def __init__(self, parent=None, name=None, workdir=None, context=None):
         elif workdir is not None:
             self._workdir_init(workdir)
 
+        # Workdir tree is used as the cwd for prepare, execute
+        # and finish steps.
+        self.workdir_tree = None
```
Comment (pvalena) [inline tmt/utils.py:130]: Why is it `None`?

File: tmt/steps/provision/__init__.py [hunk 13]
```diff
@@ -392,11 +393,12 @@ def push(self, source=None, destination=None, options=None):
...
         if destination is None:
             destination = "/"
         if source is None:
-            source = self.parent.plan.workdir
+            source = f"{self.parent.plan.workdir} " \
+                     f"{self.parent.plan.run.workdir_tree}"
```
Comment (pvalena) [inline tmt/steps/provision/__init__.py:401]: Why there're two paths? Please clarify how it is handled?
I assume rsync simply copies both paths [folders] into the destination [folder]. Do we assume that `source` has no spaces / is shell-safe?

Also, there's no check for trailing `/` which copies only the contents of the folder (`rsync` behavior). Is that expected?

Same for the `destination`, actually.

File: tmt/base.py [hunk 14]
```diff
@@ -980,6 +989,37 @@ def _save_tree(self, tree):
...
+        self.debug(
+            f"Create git worktree from '{tree_root}' to '{self.workdir_tree}'")
+        # Branch must be set, it actually names the worktree
+        branch_name = self.workdir.replace('/', '-').lstrip('-')
+        # Make sure command runs against existing an existing directory
+        self.run(
+            f'git -C {tree_root} worktree '
```
Comment (pvalena) [inline tmt/base.py:1019]: Can be simplified:
```suggestion
            f'git worktree '
```

File: tmt/base.py [hunk 15]
```diff
@@ -980,6 +989,37 @@ def _save_tree(self, tree):
...
+            f'add -b {branch_name} {self.workdir_tree}',
+            cwd=self.workdir)
```
Comment (pvalena) [inline tmt/base.py:1021]:
```suggestion
            cwd=tree_root)
```

**Review by pvalena — COMMENTED** [review]
Looks good, please clarify - see comments.

_ _ _ _

_Note: the `base.py` file has 1500+ lines now... Could we maybe split it into multiple, one `class` per file?_

File: tmt/steps/finish/__init__.py (hunk 9 above)
Comment (thrix) [reply tmt/steps/finish/__init__.py:22]: removed

File: tmt/steps/provision/__init__.py (hunk 8 above)
Comment (thrix) [reply tmt/steps/provision/__init__.py:396]: will be resolved soonish

File: tmt/base.py (hunk 15 above)
Comment (thrix) [reply tmt/base.py]: we decided to ditch `git worktree` in this release ... so this code was removed

File: tmt/base.py (hunk 14 above)
Comment (thrix) [reply tmt/base.py]: we decided to ditch `git worktree` in this release ... so this code was removed

File: tmt/steps/provision/__init__.py (hunk 13 above)
Comment (thrix) [reply tmt/steps/provision/__init__.py]: yeah, this is not important anymore, we sync only workdir as before, thus resolving.

In the reviewed patch set this tried to sync multiple directories without a common parent, thus 2 paths ... but yeah, as this is changed back now, I am resolving ...

File: tmt/utils.py (hunk 12 above)
Comment (thrix) [reply tmt/utils.py]: This is gone. Idea was to have this always initialized directly in the "top" parent class, but I changed it this is not needed anymore. The initialization is do only in `Plan` now.

**Comment by thrix** [comment]
mmkay, I need to add a few more tests in the morning, and make sure all drivers work and it should be ready

**Comment by thrix** [comment]
@lukaszachy @psss @pvalena @jscotka  this should be now ready for review

**Comment by thrix** [comment]
tried to correct the Fedora-34 problem, nice to see new fmf detecting duplicate keys \o/

File: tests/run/worktree/test.sh [hunk 16]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+. /usr/share/beakerlib/beakerlib.sh || exit 1
+
+rlJournalStart
+    rlPhaseStartSetup
+        rlRun "tmp=\$(mktemp -d)" 0 "Creating tmp directory"
+        rlRun "cp -rf data/* $tmp"
```
Comment (psss) [inline tests/run/worktree/test.sh:7]: Do we need to copy the files? The test does not store any files in the `data` directory.

File: tests/run/worktree/test.sh [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+        rlRun "set -o pipefail"
```
Comment (psss) [inline tests/run/worktree/test.sh:8]: No pipe in the test, not needed.

File: tests/run/worktree/test.sh [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+    rlPhaseEnd
+
+    rlPhaseStartTest "Simple"
+        rlRun "pushd $tmp/simple"
+        rlRun "tmt run"
+        rlRun "tmt run --last report -vvv"
```
Comment (psss) [inline tests/run/worktree/test.sh:14]: What about `tmt run -ar report -vvv` in one line and use `-r` to remove the workdir?

File: tmt/base.py [hunk 19]
```diff
@@ -430,7 +431,7 @@ class Plan(Node):
     def __init__(self, node, run=None):
         """ Initialize the plan """
         super().__init__(node, parent=run)
-        self.run = run
+        self.my_run = run
```
Comment (thrix) [inline tmt/base.py:433]: for reviewers: note that `self.run` collides with `Common.run` function .....

File: tmt/base.py [hunk 20]
```diff
@@ -465,13 +469,26 @@ def __init__(self, node, run=None):
...
             return combined
         else:
             return self._environment
 
+    def _initialize_workdir_tree(self):
+        """
+        Workdir tree, a copy of the tree root, used as cwd in prepare, execute
+        and finish steps
+        """
+        # Bail out if workdir does not exist, we are not running the plan ...
```
Comment (thrix) [inline tmt/base.py:483]:
```suggestion
        # Bail out if workdir does not exist, e.g. not running the plan ...
```

File: tmt/utils.py [hunk 21]
```diff
@@ -410,6 +410,11 @@ def run(
 
         # Run the command, handle the exit code
         cwd = cwd or self.workdir
+
+        # Fail nicely if the working directory does not exist
+        if not os.path.exists(cwd):
+            raise GeneralError(f'The working directory {cwd} does not exist')
```
Comment (psss) [inline tmt/utils.py:416]:
```suggestion
            raise RunError(f"The working directory '{cwd}' does not exist.")
```

File: tmt/base.py [hunk 22]
```diff
@@ -951,6 +966,7 @@ def __init__(self, id_=None, tree=None, context=None):
             raise tmt.utils.GeneralError(
                 "Run id has to be specified in order to use --follow.")
         super().__init__(workdir=id_ or True, context=context)
+
```
Comment (thrix) [inline tmt/base.py:969]:
```suggestion
```

File: tmt/steps/__init__.py [hunk 23]
```diff
@@ -61,7 +61,8 @@ def __init__(self, data={}, plan=None, name=None):
     def enabled(self):
         """ True if the step is enabled """
         try:
-            return self.name in self.plan.run._context.obj.steps
+            return self.name in self.plan.my_run._context.obj.steps
+
```
Comment (thrix) [inline tmt/steps/__init__.py:65]:
```suggestion
```

File: tmt/steps/provision/__init__.py [hunk 24]
```diff
@@ -396,7 +398,7 @@ def push(self, source=None, destination=None, options=None):
         if destination is None:
             destination = "/"
         if source is None:
-            source = self.parent.plan.workdir
+            source = f"{self.parent.plan.workdir}"
```
Comment (thrix) [inline tmt/steps/provision/__init__.py:401]:
```suggestion
            source = self.parent.plan.workdir
￼```

File: tmt/steps/provision/podman.py [hunk 25]
```diff
@@ -128,31 +130,38 @@ def start(self):
...
+            f'ansible-playbook '
             f'{self._ansible_verbosity()} -c podman -i {self.container}, '
-            f'{playbook}')
+            f'{playbook}',
+            cwd=self.parent.plan.workdir_tree)
         self._ansible_summary(stdout)
 
     def podman(self, command, **kwargs):
         """ Run given command via podman """
+        self.debug(' '.join(command))
```
Comment (thrix) [inline tmt/steps/provision/podman.py:164]:
```suggestion
```

File: tmt/base.py [hunk 26]
```diff
@@ -465,13 +468,26 @@ def __init__(self, node, run=None):
...
+        """
+        Workdir tree, a copy of the tree root, used as cwd in prepare, execute
+        and finish steps
+        """
+        # Bail out if workdir does not exist, e.g. not running the plan ...
+        if not self.workdir:
+            return
+
+        self.workdir_tree = os.path.join(self.workdir, 'tree')
+        tree_root = self.my_run.tree.root or '.'
```
Comment (psss) [inline tmt/base.py:488]: Why defaulting to `.`? There has to be a metadata tree root. Otherwise we would copy a random current working directory? Or did I miss anything?

File: tmt/steps/provision/__init__.py [hunk 27]
```diff
@@ -397,6 +399,7 @@ def push(self, source=None, destination=None, options=None):
             destination = "/"
         if source is None:
             source = self.parent.plan.workdir
+￼```
```
Comment (psss) [inline tmt/steps/provision/__init__.py:402]: Some leftover.

File: tmt/steps/execute/internal.py [hunk 28]
```diff
@@ -117,7 +117,6 @@ def go(self):
         for guest in self.step.plan.provision.guests():
 
             # Push workdir to guest and execute tests
```
Comment (psss) [inline tmt/steps/execute/internal.py:119]: Adjust the comment as well.

File: tmt/steps/provision/podman.py [hunk 29]
```diff
@@ -128,27 +130,33 @@ def start(self):
                 ['pull', '-q', self.image],
                 message=f"Pull image '{self.image}'.")
 
-        # Deduce container name from run id, as it can be a path,
+        # Mount the whole run directory in the container
+        workdir = self.parent.plan.workdir
+
+        # Deduce container name from workdir, bacause it is a path
```
Comment (psss) [inline tmt/steps/provision/podman.py:136]:
```suggestion
        # Deduce container name from workdir, because it is a path
```

File: tmt/steps/provision/podman.py [hunk 30]
```diff
@@ -128,27 +130,33 @@ def start(self):
                 ['pull', '-q', self.image],
                 message=f"Pull image '{self.image}'.")
 
-        # Deduce container name from run id, as it can be a path,
+        # Mount the whole run directory in the container
```
Comment (psss) [inline tmt/steps/provision/podman.py:133]:
```suggestion
        # Mount the whole plan directory in the container
```

File: tests/run/worktree/main.fmf [hunk 31]
```diff
@@ -0,0 +1,6 @@
+summary: Check whether worktree is correctly set for all steps.
+description:
+    For `prepare`, `finish` step and a simple test defined directly
+    in the execute step, the current working directory must be set
```
Comment (psss) [inline tests/run/worktree/main.fmf:4]: For simple tests defined directly in the `execute` step we still use `discover/default/tests` as cwd.

**Review by psss — CHANGES_REQUESTED** [review]
Seems to be working fine. Thanks. Added a couple of comments.

File: tmt/steps/execute/internal.py [hunk 32]
```diff
@@ -117,7 +117,6 @@ def go(self):
         for guest in self.step.plan.provision.guests():
```
Comment (thrix) [inline tmt/steps/execute/internal.py:118]:
```suggestion
```

File: tmt/steps/execute/internal.py [hunk 33]
```diff
@@ -115,9 +115,7 @@ def go(self):
         # For each guest execute all tests
         tests = self.prepare_tests()
         for guest in self.step.plan.provision.guests():
-
             # Push workdir to guest and execute tests
```
Comment (thrix) [inline tmt/steps/execute/internal.py:118]:
````
suggestion
```

File: tmt/steps/execute/internal.py (hunk 28 above)
Comment (thrix) [reply tmt/steps/execute/internal.py]: adjusted

File: tmt/steps/provision/__init__.py (hunk 27 above)
Comment (thrix) [reply tmt/steps/provision/__init__.py:402]:
```suggestion
```

File: tests/run/worktree/test.sh (hunk 16 above)
Comment (thrix) [reply tests/run/worktree/test.sh:7]: yeah, no files created, so no

File: tests/run/worktree/test.sh (hunk 17 above)
Comment (thrix) [reply tests/run/worktree/test.sh:8]: removed whole setup

File: tmt/base.py (hunk 26 above)
Comment (thrix) [reply tmt/base.py:488]: hmm, trying to remember :)

File: tests/run/worktree/main.fmf (hunk 31 above)
Comment (thrix) [reply tests/run/worktree/main.fmf:4]: hmm, @psss so I guess that is something to fix? I do not think that is correct right?

File: tmt/base.py (hunk 26 above)
Comment (thrix) [reply tmt/base.py:488]: @psss maybe bacause of this?

```
$ tmt
No metadata found in the '.' directory. Use 'tmt init' to get started.

$ tmt run -a provision -h container execute --how shell --script "echo hi"
/var/tmp/tmt/run-015

/plans/default
    warn: The 'shell' execute method has been deprecated.
    warn: Use 'how: tmt' in the execute step instead (L2).
    warn: Set 'framework: shell' in test metadata (L1).
    warn: Support for old methods will be dropped in March 2021.
    discover
        how: shell
        summary: 1 test selected
    provision
        how: container
        image: fedora
        distro: Fedora 33 (Container Image)
        summary: 1 guest provisioned
    prepare
        summary: 0 preparations applied
    execute
        how: tmt
        summary: 1 test executed
    report
        how: display
        summary: 1 test passed
    finish
        container: stopped
        container: removed
        summary: 0 tasks completed

total: 1 test passed
```

File: tests/run/worktree/main.fmf (hunk 31 above)
Comment (thrix) [reply tests/run/worktree/main.fmf:4]: right, there is nothing to fix, just adjust the description.

Comment (thrix) [reply tests/run/worktree/main.fmf]: corrected test description,

**Comment by thrix** [comment]
great, all tests pass, but we do not get back results from packit, not sure WHY

File: tmt/base.py (hunk 26 above)
Comment (psss) [reply tmt/base.py:488]: I think this could be quite dangerous: Let's say I just want to reserve a box for experimenting using:

    tmt run provision login finish

And because I just started a new shell for that I'm standing in my home directory. So this would start rsyncing my whole home directory to the workdir...

Comment (thrix) [reply tmt/base.py]: Agreed to skip copy in this case, and jsut print a debug message.

**Review by psss — APPROVED** [review]
Thanks a lot for implementing this, @thrix. I adjusted the test provision method selection and `worktree` naming as discussed on irc. There's one a little bit more important change: The `guest.push()` call cannot be removed from the executor (after updating tests and reexecuting them using `discover -f execute -f` the new test content would not be synced) and the first call should not be done before we make sure `rsync` is installed on the guest, which is during the `prepare` step. Both should be fixed. Please, have a look if this change is ok. Thanks.
